File: Comparison Programming Languages/DatabasePerformanceTest.go
package main

import (
	"database/sql"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"sync"
	"time"

	_ "github.com/lib/pq"
)

// 数据库配置
const (
	DB_URL       = "host=localhost port=5432 user=postgres password=Dr141592 dbname=project sslmode=disable"
	SCHEMA_NAME  = "test_schema"
	TABLE_NAME   = SCHEMA_NAME + ".test_perf"
	TOTAL_ROWS   = 500000
	BATCH_SIZE   = 10000
	CONCURRENT   = 10
	SAMPLE_SIZE  = 1000
	CHARACTERS   = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
)

var rnd *rand.Rand

func init() {
	// 初始化随机数生成器
	rnd = rand.New(rand.NewSource(time.Now().UnixNano()))
}

// 生成随机字符串
func generateRandomStr(length int) string {
	var sb strings.Builder
	sb.Grow(length)
	for i := 0; i < length; i++ {
		sb.WriteByte(CHARACTERS[rnd.Intn(len(CHARACTERS))])
	}
	return sb.String()
}

// 初始化Schema和表
func initSchemaAndTable(db *sql.DB) error {
	_, err := db.Exec("CREATE SCHEMA IF NOT EXISTS " + SCHEMA_NAME)
	if err != nil {
		return err
	}

	_, err = db.Exec("DROP TABLE IF EXISTS " + TABLE_NAME)
	if err != nil {
		return err
	}

	createTableSQL := fmt.Sprintf(`
		CREATE TABLE %s (
			id SERIAL PRIMARY KEY,
			uid VARCHAR(32) NOT NULL UNIQUE,
			content TEXT NOT NULL,
			value INT NOT NULL,
			create_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
		)
	`, TABLE_NAME)
	_, err = db.Exec(createTableSQL)
	if err != nil {
		return err
	}

	_, err = db.Exec(fmt.Sprintf(
		"CREATE INDEX idx_%s_test_perf_value ON %s(value)",
		SCHEMA_NAME, TABLE_NAME))
	if err != nil {
		return err
	}

	fmt.Printf("Schema '%s'和表 '%s' 创建完成\n", SCHEMA_NAME, TABLE_NAME)
	return nil
}

// 批量插入数据
func insertLargeData(db *sql.DB, totalRows, batchSize int) error {
	fmt.Printf("开始插入 %d 条数据...\n", totalRows)
	totalBatches := (totalRows + batchSize - 1) / batchSize

	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(fmt.Sprintf(
		"INSERT INTO %s (uid, content, value) VALUES ($1, $2, $3)",
		TABLE_NAME))
	if err != nil {
		return err
	}
	defer stmt.Close()

	for batch := 0; batch < totalBatches; batch++ {
		currentBatchSize := batchSize
		if remaining := totalRows - batch*batchSize; remaining < batchSize {
			currentBatchSize = remaining
		}

		for i := 0; i < currentBatchSize; i++ {
			uid := generateRandomStr(32)
			content := generateRandomStr(100)
			value := rnd.Intn(1000) + 1

			_, err := stmt.Exec(uid, content, value)
			if err != nil {
				return err
			}
		}

		if err := tx.Commit(); err != nil {
			return err
		}

		// 为下一批次准备新事务
		tx, err = db.Begin()
		if err != nil {
			return err
		}
		stmt, err = tx.Prepare(fmt.Sprintf(
			"INSERT INTO %s (uid, content, value) VALUES ($1, $2, $3)",
			TABLE_NAME))
		if err != nil {
			return err
		}

		if (batch+1)%10 == 0 {
			progress := float64(batch+1) / float64(totalBatches) * 100
			fmt.Printf("插入进度: %.1f%% (%d/%d批次)\n", progress, batch+1, totalBatches)
		}
	}

	// 提交最后一个事务
	tx.Commit()
	return nil
}

// 获取最大ID
func getMaxId(db *sql.DB) (int, error) {
	var maxId int
	err := db.QueryRow("SELECT MAX(id) FROM " + TABLE_NAME).Scan(&maxId)
	return maxId, err
}

// 测试查询性能
func testSelectPerformance(db *sql.DB) error {
	maxId, err := getMaxId(db)
	if err != nil {
		return err
	}
	if maxId == 0 {
		return nil
	}

	// 单条主键查询
	var totalSingle int64
	stmt, err := db.Prepare(fmt.Sprintf("SELECT * FROM %s WHERE id = $1", TABLE_NAME))
	if err != nil {
		return err
	}
	defer stmt.Close()

	for i := 0; i < SAMPLE_SIZE; i++ {
		id := rnd.Intn(maxId) + 1
		start := time.Now()
		rows, err := stmt.Query(id)
		if err != nil {
			return err
		}
		rows.Close() // 立即关闭结果集
		totalSingle += time.Since(start).Nanoseconds()
	}
	avgSingle := float64(totalSingle) / 1e9 / float64(SAMPLE_SIZE)

	// 条件查询
	var totalCond int64
	condStmt, err := db.Prepare(fmt.Sprintf("SELECT * FROM %s WHERE value = $1 LIMIT 10", TABLE_NAME))
	if err != nil {
		return err
	}
	defer condStmt.Close()

	for i := 0; i < SAMPLE_SIZE; i++ {
		value := rnd.Intn(1000) + 1
		start := time.Now()
		rows, err := condStmt.Query(value)
		if err != nil {
			return err
		}
		rows.Close()
		totalCond += time.Since(start).Nanoseconds()
	}
	avgCond := float64(totalCond) / 1e9 / float64(SAMPLE_SIZE)

	// 范围查询
	rangeStmt, err := db.Prepare(fmt.Sprintf(
		"SELECT * FROM %s WHERE value BETWEEN $1 AND $2 ORDER BY create_time LIMIT 1000",
		TABLE_NAME))
	if err != nil {
		return err
	}
	defer rangeStmt.Close()

	start := time.Now()
	rows, err := rangeStmt.Query(400, 600)
	if err != nil {
		return err
	}
	rows.Close()
	rangeSec := time.Since(start).Seconds()

	fmt.Println("查询性能:")
	fmt.Printf("  单条主键查询（%d次平均）: %.6f秒/次\n", SAMPLE_SIZE, avgSingle)
	fmt.Printf("  条件查询（%d次平均）: %.6f秒/次\n", SAMPLE_SIZE, avgCond)
	fmt.Printf("  范围查询（1000条结果）: %.6f秒\n", rangeSec)

	return nil
}

// 测试更新性能
func testUpdatePerformance(db *sql.DB) error {
	maxId, err := getMaxId(db)
	if err != nil {
		return err
	}
	if maxId == 0 {
		return nil
	}

	var totalTime int64
	stmt, err := db.Prepare(fmt.Sprintf("UPDATE %s SET value = $1 WHERE id = $2", TABLE_NAME))
	if err != nil {
		return err
	}
	defer stmt.Close()

	for i := 0; i < SAMPLE_SIZE; i++ {
		targetId := rnd.Intn(maxId) + 1
		newValue := rnd.Intn(1000) + 1

		start := time.Now()
		_, err := stmt.Exec(newValue, targetId)
		if err != nil {
			return err
		}
		totalTime += time.Since(start).Nanoseconds()
	}

	avgUpdate := float64(totalTime) / 1e9 / float64(SAMPLE_SIZE)
	fmt.Printf("更新性能（%d次平均）: %.6f秒/次\n", SAMPLE_SIZE, avgUpdate)
	return nil
}

// 测试删除性能
func testDeletePerformance(db *sql.DB) error {
	// 备份数据
	type backupRow struct {
		id      int
		uid     string
		content string
		value   int
	}
	var backup []backupRow

	rows, err := db.Query(fmt.Sprintf(
		"SELECT id, uid, content, value FROM %s ORDER BY random() LIMIT %d",
		TABLE_NAME, SAMPLE_SIZE))
	if err != nil {
		return err
	}
	defer rows.Close()

	for rows.Next() {
		var row backupRow
		if err := rows.Scan(&row.id, &row.uid, &row.content, &row.value); err != nil {
			return err
		}
		backup = append(backup, row)
	}

	if len(backup) == 0 {
		return nil
	}

	// 执行删除
	var totalTime int64
	deleteStmt, err := db.Prepare(fmt.Sprintf("DELETE FROM %s WHERE id = $1", TABLE_NAME))
	if err != nil {
		return err
	}
	defer deleteStmt.Close()

	for _, row := range backup {
		start := time.Now()
		_, err := deleteStmt.Exec(row.id)
		if err != nil {
			return err
		}
		totalTime += time.Since(start).Nanoseconds()
	}

	// 恢复数据
	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	insertStmt, err := tx.Prepare(fmt.Sprintf(
		"INSERT INTO %s (id, uid, content, value) VALUES ($1, $2, $3, $4)",
		TABLE_NAME))
	if err != nil {
		return err
	}
	defer insertStmt.Close()

	for _, row := range backup {
		_, err := insertStmt.Exec(row.id, row.uid, row.content, row.value)
		if err != nil {
			return err
		}
	}

	if err := tx.Commit(); err != nil {
		return err
	}

	avgDelete := float64(totalTime) / 1e9 / float64(SAMPLE_SIZE)
	fmt.Printf("删除性能（%d次平均）: %.6f秒/次\n", SAMPLE_SIZE, avgDelete)
	return nil
}

// 并发任务结果
type taskResult struct {
	id     int
	action string
	time   float64
	error  string
}

// 处理查询操作
func handleSelect(db *sql.DB) error {
	maxId, err := getMaxId(db)
	if err != nil || maxId <= 0 {
		return err
	}

	id := rnd.Intn(maxId) + 1
	stmt, err := db.Prepare(fmt.Sprintf("SELECT * FROM %s WHERE id = $1", TABLE_NAME))
	if err != nil {
		return err
	}
	defer stmt.Close()

	rows, err := stmt.Query(id)
	if err != nil {
		return err
	}
	rows.Close()
	return nil
}

// 处理更新操作
func handleUpdate(db *sql.DB) error {
	maxId, err := getMaxId(db)
	if err != nil || maxId <= 0 {
		return err
	}

	targetId := rnd.Intn(maxId) + 1
	newValue := rnd.Intn(1000) + 1

	stmt, err := db.Prepare(fmt.Sprintf("UPDATE %s SET value = $1 WHERE id = $2", TABLE_NAME))
	if err != nil {
		return err
	}
	defer stmt.Close()

	_, err = stmt.Exec(newValue, targetId)
	return err
}

// 处理删除恢复操作
func handleDeleteRestore(db *sql.DB) error {
	var id int
	var uid, content string
	var value int

	err := db.QueryRow(fmt.Sprintf(
		"SELECT id, uid, content, value FROM %s ORDER BY random() LIMIT 1",
		TABLE_NAME)).Scan(&id, &uid, &content, &value)
	if err != nil {
		return err
	}

	// 删除
	_, err = db.Exec(fmt.Sprintf("DELETE FROM %s WHERE id = $1", TABLE_NAME), id)
	if err != nil {
		return err
	}

	// 恢复
	_, err = db.Exec(fmt.Sprintf(
		"INSERT INTO %s (id, uid, content, value) VALUES ($1, $2, $3, $4)",
		TABLE_NAME), id, uid, content, value)
	return err
}

// 并发任务执行函数
func concurrentTask(id int, results chan<- taskResult) {
	result := taskResult{id: id}
	db, err := sql.Open("postgres", DB_URL)
	if err != nil {
		result.action = "error"
		result.error = err.Error()
		results <- result
		return
	}
	defer db.Close()

	// 随机选择操作类型
	var action string
	switch rnd.Intn(3) {
	case 0:
		action = "select"
	case 1:
		action = "update"
	default:
		action = "delete_restore"
	}

	start := time.Now()
	var execErr error

	switch action {
	case "select":
		execErr = handleSelect(db)
	case "update":
		execErr = handleUpdate(db)
	case "delete_restore":
		execErr = handleDeleteRestore(db)
	}

	if execErr != nil {
		result.action = "error"
		result.error = execErr.Error()
	} else {
		result.action = action
		result.time = time.Since(start).Seconds()
	}

	results <- result
}

// 测试并发性能
func testConcurrentPerformance() {
	totalOps := 1000
	fmt.Printf("开始并发测试（%d线程，共%d次操作）...\n", CONCURRENT, totalOps)

	start := time.Now()
	results := make(chan taskResult, totalOps)
	var wg sync.WaitGroup

	// 启动工作池
	pool := make(chan struct{}, CONCURRENT)
	for i := 0; i < totalOps; i++ {
		pool <- struct{}{}
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			defer func() { <-pool }()
			concurrentTask(id, results)
		}(i)
	}

	// 等待所有任务完成并关闭通道
	go func() {
		wg.Wait()
		close(results)
	}()

	// 收集结果
	success := 0
	var errors []string
	actionTimes := map[string][]float64{
		"select":         {},
		"update":         {},
		"delete_restore": {},
	}

	for res := range results {
		if res.action == "error" {
			errors = append(errors, res.error)
		} else {
			success++
			actionTimes[res.action] = append(actionTimes[res.action], res.time)
		}
	}

	totalTime := time.Since(start).Seconds()

	// 计算平均时间
	avgTimes := make(map[string]float64)
	for action, times := range actionTimes {
		if len(times) == 0 {
			avgTimes[action] = 0
			continue
		}
		var sum float64
		for _, t := range times {
			sum += t
		}
		avgTimes[action] = sum / float64(len(times))
	}

	fmt.Println("并发测试完成:")
	fmt.Printf("  总耗时: %.6f秒\n", totalTime)
	fmt.Printf("  总操作数: %d，成功: %d，失败: %d\n", totalOps, success, len(errors))
	fmt.Println("  平均耗时（按操作类型）:")
	fmt.Printf("    查询: %.6f秒/次\n", avgTimes["select"])
	fmt.Printf("    更新: %.6f秒/次\n", avgTimes["update"])
	fmt.Printf("    删除恢复: %.6f秒/次\n", avgTimes["delete_restore"])

	if len(errors) > 0 {
		fmt.Printf("  错误示例: %s\n", errors[0])
	}
}

func main() {
	// 连接数据库
	db, err := sql.Open("postgres", DB_URL)
	if err != nil {
		fmt.Printf("无法连接数据库: %v\n", err)
		os.Exit(1)
	}
	defer db.Close()

	// 测试连接
	if err := db.Ping(); err != nil {
		fmt.Printf("数据库连接失败: %v\n", err)
		os.Exit(1)
	}

	// 1. 初始化Schema和表
	fmt.Println("\n===== 1. 初始化Schema和表 =====")
	initStart := time.Now()
	if err := initSchemaAndTable(db); err != nil {
		fmt.Printf("初始化失败: %v\n", err)
		return
	}
	initTime := time.Since(initStart).Seconds()
	fmt.Printf("初始化耗时: %.6f秒\n", initTime)

	// 2. 插入大量数据
	fmt.Println("\n===== 2. 插入大量数据 =====")
	insertStart := time.Now()
	if err := insertLargeData(db, TOTAL_ROWS, BATCH_SIZE); err != nil {
		fmt.Printf("插入数据失败: %v\n", err)
		return
	}
	insertTime := time.Since(insertStart).Seconds()
	fmt.Printf("插入%d万条数据总耗时: %.6f秒，平均每条: %.8f秒\n",
		TOTAL_ROWS/10000, insertTime, insertTime/float64(TOTAL_ROWS))

	// 3. 测试查询性能
	fmt.Println("\n===== 3. 测试查询性能 =====")
	selectStart := time.Now()
	if err := testSelectPerformance(db); err != nil {
		fmt.Printf("查询测试失败: %v\n", err)
		return
	}
	selectTime := time.Since(selectStart).Seconds()
	fmt.Printf("查询测试总耗时: %.6f秒\n", selectTime)

	// 4. 测试更新性能
	fmt.Println("\n===== 4. 测试更新性能 =====")
	updateStart := time.Now()
	if err := testUpdatePerformance(db); err != nil {
		fmt.Printf("更新测试失败: %v\n", err)
		return
	}
	updateTime := time.Since(updateStart).Seconds()
	fmt.Printf("更新测试总耗时: %.6f秒\n", updateTime)

	// 5. 测试删除性能
	fmt.Println("\n===== 5. 测试删除性能 =====")
	deleteStart := time.Now()
	if err := testDeletePerformance(db); err != nil {
		fmt.Printf("删除测试失败: %v\n", err)
		return
	}
	deleteTime := time.Since(deleteStart).Seconds()
	fmt.Printf("删除测试总耗时: %.6f秒\n", deleteTime)

	// 6. 测试并发性能
	fmt.Println("\n===== 6. 测试并发性能 =====")
	testConcurrentPerformance()

	fmt.Println("\n所有测试完成，资源已释放")
}

File: Comparison Programming Languages/DatabasePerformanceTest.py
import psycopg2
import time
import random
import string
import os
from psycopg2 import pool
from psycopg2.extras import execute_batch
import concurrent.futures

DB_CONFIG = {
    "dbname": "sustc_db",
    "user": "postgres",
    "password": "676767",
    "host": "localhost",
    "port": "5432"
}
SCHEMA_NAME = "test_schema"
TABLE_NAME = f"{SCHEMA_NAME}.test_perf"
TOTAL_ROWS = 500000
BATCH_SIZE = 10000
CONCURRENT_WORKERS = 10


def generate_random_str(length=10):
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))


def time_it(func, *args, **kwargs):
    start = time.perf_counter()
    result = func(*args, **kwargs)
    end = time.perf_counter()
    return result, end - start


def init_schema_and_table(cursor):
    cursor.execute(f"CREATE SCHEMA IF NOT EXISTS {SCHEMA_NAME}")
    cursor.execute(f"""
        DROP TABLE IF EXISTS {TABLE_NAME};
        CREATE TABLE {TABLE_NAME} (
            id SERIAL PRIMARY KEY,
            uid VARCHAR(32) NOT NULL UNIQUE,
            content TEXT NOT NULL,
            value INT NOT NULL,
            create_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );

        CREATE INDEX idx_{SCHEMA_NAME}_test_perf_value ON {TABLE_NAME}(value);
    """)
    print(f"Schema '{SCHEMA_NAME}'和表 '{TABLE_NAME}' 创建完成")


def insert_large_data(cursor, total_rows, batch_size):
    print(f"开始插入 {total_rows} 条数据...")
    total_batches = (total_rows + batch_size - 1) // batch_size

    for batch in range(total_batches):
        data = [
            (
                generate_random_str(32),
                generate_random_str(100),
                random.randint(1, 1000)
            )
            for _ in range(batch_size) if (batch * batch_size + _) < total_rows
        ]

        execute_batch(cursor, f"""
            INSERT INTO {TABLE_NAME} (uid, content, value)
            VALUES (%s, %s, %s)
        """, data, page_size=1000)

        if (batch + 1) % 10 == 0:
            progress = (batch + 1) / total_batches * 100
            print(f"插入进度: {progress:.1f}% ({batch + 1}/{total_batches}批次)")


def test_select_performance(cursor, sample_size=1000):
    cursor.execute(f"SELECT MAX(id) FROM {TABLE_NAME}")
    max_id = cursor.fetchone()[0]
    if not max_id:
        return

    total_single = 0
    for _ in range(sample_size):
        _, t = time_it(
            cursor.execute,
            f"SELECT * FROM {TABLE_NAME} WHERE id = %s",
            (random.randint(1, max_id),)
        )
        total_single += t
    avg_single = total_single / sample_size

    total_cond = 0
    for _ in range(sample_size):
        val = random.randint(1, 1000)
        _, t = time_it(
            cursor.execute,
            f"SELECT * FROM {TABLE_NAME} WHERE value = %s LIMIT 10",
            (val,)
        )
        total_cond += t
    avg_cond = total_cond / sample_size

    _, t_range = time_it(
        cursor.execute,
        f"SELECT * FROM {TABLE_NAME} WHERE value BETWEEN %s AND %s ORDER BY create_time LIMIT 1000",
        (400, 600)
    )

    print(f"查询性能:")
    print(f"  单条主键查询（{sample_size}次平均）: {avg_single:.6f}秒/次")
    print(f"  条件查询（{sample_size}次平均）: {avg_cond:.6f}秒/次")
    print(f"  范围查询（1000条结果）: {t_range:.6f}秒")


def test_update_performance(cursor, sample_size=1000):
    cursor.execute(f"SELECT MAX(id) FROM {TABLE_NAME}")
    max_id = cursor.fetchone()[0]
    if not max_id:
        return

    total_time = 0
    for _ in range(sample_size):
        target_id = random.randint(1, max_id)
        new_val = random.randint(1, 1000)
        _, t = time_it(
            cursor.execute,
            f"UPDATE {TABLE_NAME} SET value = %s WHERE id = %s",
            (new_val, target_id)
        )
        total_time += t

    avg_update = total_time / sample_size
    print(f"更新性能（{sample_size}次平均）: {avg_update:.6f}秒/次")


def test_delete_performance(cursor, sample_size=1000):
    cursor.execute(f"""
        SELECT id, uid, content, value FROM {TABLE_NAME} 
        ORDER BY random() LIMIT {sample_size}
    """)
    backup = cursor.fetchall()
    if not backup:
        return

    total_time = 0
    deleted_ids = [row[0] for row in backup]
    for idx in deleted_ids:
        _, t = time_it(
            cursor.execute,
            f"DELETE FROM {TABLE_NAME} WHERE id = %s",
            (idx,)
        )
        total_time += t

    execute_batch(cursor, f"""
        INSERT INTO {TABLE_NAME} (id, uid, content, value)
        VALUES (%s, %s, %s, %s)
    """, backup)

    avg_delete = total_time / sample_size
    print(f"删除性能（{sample_size}次平均）: {avg_delete:.6f}秒/次")


def concurrent_operation(operation_id):
    conn = None
    cursor = None
    try:
        conn = psycopg2.connect(**DB_CONFIG)
        conn.autocommit = True
        cursor = conn.cursor()

        # 30% select, 30% update, 40% delete_restore
        p = random.random()
        if p < 0.3:
            action = "select"
        elif p < 0.6:
            action = "update"
        else:
            action = "delete_restore"

        start = time.perf_counter()

        if action == "select":
            cursor.execute(f"SELECT MAX(id) FROM {TABLE_NAME}")
            max_id = cursor.fetchone()[0]
            if max_id:
                cursor.execute(f"SELECT * FROM {TABLE_NAME} WHERE id = %s", (random.randint(1, max_id),))
                cursor.fetchone()     # ⭐ fetch result! 重要！

        elif action == "update":
            cursor.execute(f"SELECT MAX(id) FROM {TABLE_NAME}")
            max_id = cursor.fetchone()[0]
            if max_id:
                cursor.execute(
                    f"UPDATE {TABLE_NAME} SET value = %s WHERE id = %s",
                    (random.randint(1, 1000), random.randint(1, max_id))
                )

        elif action == "delete_restore":
            cursor.execute(f"SELECT id, uid, content, value FROM {TABLE_NAME} ORDER BY random() LIMIT 1")
            row = cursor.fetchone()
            if row:
                cursor.execute(f"DELETE FROM {TABLE_NAME} WHERE id = %s", (row[0],))
                cursor.execute(
                    f"INSERT INTO {TABLE_NAME} (id, uid, content, value) VALUES (%s, %s, %s, %s)",
                    row
                )

        end = time.perf_counter()
        return (operation_id, action, end - start)

    except Exception as e:
        return (operation_id, "error", str(e))

    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


def test_concurrent_performance(workers=CONCURRENT_WORKERS, total_ops=1000):
    print(f"\n开始并发测试（{workers}线程，共{total_ops}次操作）...")
    start = time.perf_counter()

    with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as executor:
        futures = [executor.submit(concurrent_operation, i) for i in range(total_ops)]
        results = [f.result() for f in concurrent.futures.as_completed(futures)]

    total_time = time.perf_counter() - start

    success = 0
    errors = []
    action_times = {"select": [], "update": [], "delete_restore": []}
    for res in results:
        op_id, action, val = res
        if action == "error":
            errors.append(val)
        else:
            success += 1
            action_times[action].append(val)

    avg_times = {k: sum(v) / len(v) if v else 0 for k, v in action_times.items()}

    print(f"并发测试完成:")
    print(f"  总耗时: {total_time:.6f}秒")
    print(f"  总操作数: {total_ops}，成功: {success}，失败: {len(errors)}")
    print(f"  平均耗时（按操作类型）:")
    print(f"    查询: {avg_times['select']:.6f}秒/次")
    print(f"    更新: {avg_times['update']:.6f}秒/次")
    print(f"    删除恢复: {avg_times['delete_restore']:.6f}秒/次")
    if errors:
        print(f"  错误示例: {errors[0]}")


def main():
    connection_pool = None
    try:
        connection_pool = pool.SimpleConnectionPool(
            minconn=1,
            maxconn=5,
            **DB_CONFIG
        )
        conn = connection_pool.getconn()
        conn.autocommit = True
        cursor = conn.cursor()

        print("\n===== 1. 初始化Schema和表 =====")
        _, t_init = time_it(init_schema_and_table, cursor)
        print(f"初始化耗时: {t_init:.6f}秒")

        print("\n===== 2. 插入大量数据 =====")
        _, t_insert = time_it(insert_large_data, cursor, TOTAL_ROWS, BATCH_SIZE)
        print(f"插入{50}万条数据总耗时: {t_insert:.6f}秒，"
              f"平均每条: {t_insert / TOTAL_ROWS:.8f}秒")

        print("\n===== 3. 测试查询性能 =====")
        _, t_select = time_it(test_select_performance, cursor)
        print(f"查询测试总耗时: {t_select:.6f}秒")

        print("\n===== 4. 测试更新性能 =====")
        _, t_update = time_it(test_update_performance, cursor)
        print(f"更新测试总耗时: {t_update:.6f}秒")

        print("\n===== 5. 测试删除性能 =====")
        _, t_delete = time_it(test_delete_performance, cursor)
        print(f"删除测试总耗时: {t_delete:.6f}秒")

        print("\n===== 6. 测试并发性能 =====")
        test_concurrent_performance()

    except Exception as e:
        print(f"测试出错: {str(e)}")
    finally:
        if 'cursor' in locals():
            cursor.close()
        if 'conn' in locals() and connection_pool:
            connection_pool.putconn(conn)
        if connection_pool:
            connection_pool.closeall()
        print("\n所有测试完成，资源已释放")


if __name__ == "__main__":
    main()